src/method.jl

```diff
@@ -58,24 +58,61 @@ function argument_types(sig)
     return parameters(sig)[2:end]
 end
 
-name_of_type(x) = x
+module DummyThatHasOnlyDefaultImports end  # for working out visability
+
+function name_of_module(m::Module)
+    if Base.is_root_module(m)
+        return nameof(m)
+    else
+        return :($(name_of_module(parentmodule(m))).$(nameof(m)))
+    end
+end
+function name_of_type(x::Core.TypeName)
+    #TODO: could let user pass this in, then we could be using what is inscope for them
+    # but this is not important as we will give a correct (if overly verbose) output as is.
+    from = DummyThatHasOnlyDefaultImports
+    if Base.isvisible(x.name, x.module, from)  # avoid qualifying things that are in scope
+        return x.name
+    else
+        return :($(name_of_module(x.module)).$(x.name))
+    end
+end
+
+name_of_type(x::Symbol) = x  # Literal
+function name_of_type(x::T) where T  # Literal
+    # it is a bug in our implementation if this error every gets hit.
+    isbits(x) || throw(DomainError((x, T), "not a valid type-param"))
+    return x
+end
 name_of_type(tv::TypeVar) = tv.name
 function name_of_type(x::DataType)
-    name_sym = Symbol(x.name)
-    if isempty(x.parameters)
-        return name_sym
+    name = name_of_type(x.name)
+    # because tuples are varadic in number of type parameters having no parameters does not
+    # mean you should not write the `{}`, so we special case them here.
+    if isempty(x.parameters) && x != Tuple{}
+        return name
     else
-        parameter_names = name_of_type.(x.parameters)
-        return :($(name_sym){$(parameter_names...)})
+        parameter_names = map(name_of_type, x.parameters)
+        return :($(name){$(parameter_names...)})
     end
 end
+
+
 function name_of_type(x::UnionAll)
-    name = name_of_type(x.body)
-    whereparam = where_parameters(x.var)
-    return :($name where $whereparam)
+    # we do nested union all unwrapping so we can make the more compact:
+    # `foo{T,A} where {T, A}`` rather than the longer: `(foo{T,A} where T) where A`
+    where_params = []
+    while x isa UnionAll
+        push!(where_params, where_parameter1(x.var))
+        x = x.body
+    end
+
+    name = name_of_type(x)
+    return :($name where {$(where_params...)})
```

Comment:
This doens;'t change the semantics of the output,
but does do the more compact grouping mentioned in the comments.